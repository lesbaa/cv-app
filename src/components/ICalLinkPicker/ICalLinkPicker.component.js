/* global alert */
import React, { Component } from 'react'
import format from 'date-fns/format'
import addDays from 'date-fns/add_days'
import { compose } from 'ramda'
import styles from './ICalLinkPicker.styles'

class ICalLinkPicker extends Component {
  state = {
    url: null,
    datetime: null,
    hasError: false,
  }

  handleChange = ({ target: { value } }) => {
    this.setState({
      datetime: value,
    }, this.buildUrl)
  }

  buildUrl = () => {
    const {
      datetime,
    } = this.state

    this.setState({
      url: `/calender?datetime=${datetime}`,
    })
  }

  handleSubmit = (e) => {
    if (!this.state.url) {
      e.preventDefault()
      this.setState({
        hasError: true,
      })
      alert('Please enter a valid date / time')
    }
  }

  getTomorrow = () => compose(
    datetime => format(datetime, 'YYYY-MM-DD'),
    datetime => addDays(datetime, 1),
  )(new Date())

  render = () => {
    const inputClassNames = [
      'input',
      this.state.hasError && 'error',
    ]
      .filter(Boolean)
      .join(' ')

    return (
      <div className="ICalLinkPicker">
        <input
          className={inputClassNames}
          type="datetime-local"
          id="datetime"
          name="trip"
          onChange={this.handleChange}
        />
        <a
          className="submit"
          href={this.state.url}
          target="_blank"
          onClick={this.handleSubmit}
        >
          Remind me!
        </a>
        <style jsx>{styles}</style>
      </div>
    )
  }
}

export default ICalLinkPicker
