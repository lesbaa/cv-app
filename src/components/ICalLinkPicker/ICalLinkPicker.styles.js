import css from 'styled-jsx/css'

export default css`
.ICalLinkPicker, .ICalLinkPicker > * {
  text-align: center;
  background-blend-mode: color-burn;
  mix-blend-mode: color-burn;
}

.input {
  background: none;
  box-sizing: border-box;
  border: 1px solid #444;
  display: inline-block;
  width: 100%;
  padding: 0.25em;
  margin: 0;
  font-size: 1.2em;
}

.error {
  border: 1px solid #744;
}

.submit {
  display: block;
  border-radius: 0.25em;
  color: #eee;
  background-color: #444;
  margin: 1em 0;
  padding: 1em;
  text-align: center;
}
`
