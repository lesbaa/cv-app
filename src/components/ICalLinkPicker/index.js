import ICalLinkPicker from './ICalLinkPicker.component'

export default ICalLinkPicker
