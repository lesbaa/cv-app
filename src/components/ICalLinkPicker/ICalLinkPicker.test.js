/* eslint-env jest */
/* global window */
import React from 'react'
import { mount } from 'enzyme'

import ICalLinkPicker, { buildUrl } from './ICalLinkPicker.component'

describe('ICalLinkPicker', () => {
  const wrapper = mount(
    <ICalLinkPicker
      message="message 1"
      isVisible
    >
      <div className="child" />
    </ICalLinkPicker>
  )

  it('matches its snapshots', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('updates the state correctly on input change', () => {
    wrapper
      .find('input')
      .simulate('change', { target: { value: 'a_value' } })

    expect(wrapper.state('url')).toBe('/calender?datetime=a_value')
  })

  it('it throws an alert and sets state.hasError to true if an invalid url state field is entered', () => {
    const preventDefault = jest.fn()

    wrapper.setState({
      url: null,
    })

    window.alert = jest.fn()

    wrapper
      .find('.submit')
      .simulate('click', { preventDefault })

    expect(preventDefault).toHaveBeenCalled()
    expect(window.alert).toHaveBeenCalled()
  })

  it('follows the default behaviour is state.url is valid', () => {
    const preventDefault = jest.fn()

    wrapper.setState({
      url: '/calender?datetime=hiyahiyahiya',
    })

    window.alert = jest.fn()

    wrapper
      .find('.submit')
      .simulate('click', { preventDefault })

    expect(preventDefault).not.toHaveBeenCalled()
    expect(window.alert).not.toHaveBeenCalled()
  })
})

describe('buildUrl', () => {
  it('builds a url with the correct structure', () => {
    expect(buildUrl('hiyahiyahiya'))
      .toBe('/calender?datetime=hiyahiyahiya')
  })
})
