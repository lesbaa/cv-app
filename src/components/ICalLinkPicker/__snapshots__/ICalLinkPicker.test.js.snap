// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ICalLinkPicker matches its snapshots 1`] = `ReactWrapper {}`;
